desci-contracts/contracts/DpidAliasRegistry.sol

```diff
@@ -0,0 +1,159 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+contract DpidAliasRegistry is OwnableUpgradeable {
+    // Only written at time of initialization
+    uint256 public firstDpid;
+
+    // Incremented on each dPID mint
+    uint256 public nextDpid;
+
+    // dpid => codex streamID (resolve dPID)
+    mapping(uint256 => string) public registry;
+
+    // codex streamID => dpid (check for existing aliases)
+    mapping(string => uint256) public reverseRegistry;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function __DpidAliasRegistry_init(uint256 _firstDpid) public initializer {
+        OwnableUpgradeable.__Ownable_init();
+        firstDpid = _firstDpid;
+        nextDpid = _firstDpid;
+    }
+
+    /**
+     * Resolve the codex stream ID of a given dPID
+     * @param dpid the alias to resolve
+     */
+    function resolve(uint256 dpid) public view returns(string memory) {
+        return registry[dpid];
+    }
+
+    /**
+     * Find the dPID of a given codex stream ID, if it exists
+     * @param streamId the codex stream ID to search for
+     */
+    function find(string calldata streamId) public view returns(uint256) {
+        return reverseRegistry[streamId];
+    }
+
+    event DpidMinted (
+        uint256 dpid,
+        string streamID
+    );
+
+    /**
+     * Claim the next free dPID by pointing it to a codex stream ID
+     *
+     * @param streamId the codex stream ID to alias
+     */
+    function mintDpid(string calldata streamId) public returns(uint256) {
+        require(reverseRegistry[streamId] == 0, "stream already has a dPID");
+
+        uint256 thisDpid = nextDpid;
+
+        // map this dPID to the passed stream ID
+        registry[thisDpid] = streamId;
+
+        // map the passed stream ID to this dPID
+        reverseRegistry[streamId] = thisDpid;
+
+        emit DpidMinted(thisDpid, streamId);
```

Comment:
I think emitting an event prior to any effects is recommended, but I'm not sure how to test a re-entrancy scenario here

https://detectors.auditbase.com/reentrancy-causing-out-of-order-event-emission-solidity